# NYC DCP Historical Liquidation Project Sprint 2 Work

## TOC:
- [Setup](#first-bullet)
- [Group Checkbook NYC](#second-bullet)
- [Assigning categories based on `Budget Code` and `Contract Purpose](#third-bullet)
- [## Merge Checkbook NYC onto CPDB geometries](#fourth-bullet)

## Setup <a class="anchor" id="first-bullet"></a>

In [2]:
import pandas as pd
import numpy as np
import re
import datetime
import time
import geopandas as gpd
import geodatasets
import contextily as cx
import matplotlib.pyplot as plt
import matplotlib.font_manager
from matplotlib import ticker
import seaborn as sns
import os
import time
import sqlalchemy
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

In [3]:
# reading in data

# --- checkbook nyc data
file_names = ['../data/checkbooknyc/Citywide-Agencies/spending_29351_05312023_040004_part_0.csv', 
'../data/checkbooknyc/Citywide-Agencies/spending_29351_05312023_040004_part_1.csv',
'../data/checkbooknyc/Citywide-Agencies/spending_29351_05312023_040004_part_2.csv',
'../data/checkbooknyc/Citywide-Agencies/spending_29351_05312023_040004_part_3.csv']
df0 = pd.read_csv(file_names[0])
df1 = pd.read_csv(file_names[1])
df2 = pd.read_csv(file_names[2])
df3 = pd.read_csv(file_names[3])
frames = [df0, df1, df2, df3]
checkbook = pd.concat(frames)

# --- cpdb geometries

path = "../data/cpdb/"
cpdb_file_list = os.listdir(path)
cpdb_file_list.remove('cpdb_projects.csv')
cpdb_file_list.remove('.DS_Store')


In [4]:
# merge cpdb data 
def merge_all_cpdb_geoms(path, file_list):
    """
    input: path and list of file names
    output: GeoDataFrame
    """
    file_list = sorted(file_list, key=lambda x: int(re.search(r'\d+$',x).group()), reverse=True)

    gdf_list = []
    for f in file_list:
        gdf = gpd.read_file(path + f)
        gdf_list.append(gdf)

    all_cpdb_geoms = pd.concat(gdf_list)
    # deal with duplicates
    all_cpdb_geoms.drop_duplicates(subset='maprojid', keep='first', inplace=True, ignore_index=True)
    return all_cpdb_geoms

# joining all of Checkbook NYC onto all cpdb
def join_checkbook_geoms(df, cpdb_geoms):
    """
    input:
    df -> dataframe to be left joined onto, in this case checkbook nyc
    cpdb_geoms -> geodataframe to join onto df, contains all shape files 2017-2023
    output:
    gdf -> geodataframe of cpdb geometries left joined onto checkbook nyc
    """
    merged = df.merge(cpdb_geoms, how='left', left_on='FMS ID', right_on='maprojid', indicator=True) # left join so we can easily count how many projects did not map to a geometry
    ### may need to uncomment this later because it might break something 
    #merged.drop('Unnamed: 0', axis=1, inplace=True)
    gdf = gpd.GeoDataFrame(merged, geometry='geometry')
    return gdf

In [5]:
# merge cpdb geometries
all_cpdb = merge_all_cpdb_geoms(path, cpdb_file_list)

In [6]:
#  data cleaning 

# --- checkbook nyc data cleaning 
checkbook = checkbook[checkbook['Check Amount']<99000000] # exclude checks over 99 mil
checkbook = checkbook[checkbook['Check Amount']>=0] # exclude negative checks, include 0 checks
checkbook['FMS ID'] = checkbook['Capital Project'].str.replace(r'\s*\d+$','') # new column for joining on CPDB

/var/folders/7p/9414x_x93rvb4p6kc2ny_wl00000gn/T/ipykernel_77067/3860368422.py:6: FutureWarning: The default value of regex will change from True to False in a future version.
  checkbook['FMS ID'] = checkbook['Capital Project'].str.replace(r'\s*\d+$','') # new column for joining on CPDB


In [7]:
checkbook['FMS ID'].nunique()

16687

## Group Checkbook NYC <a class="anchor" id="second-bullet"></a>

In [8]:
# --- groupby 
temp_cols_for_grouping = ['FMS ID']
temp_cols_for_limiting = temp_cols_for_grouping + ['Contract Purpose', 'Agency', 'Budget Code', 'Check Amount']

df_limited_cols = checkbook.loc[:, temp_cols_for_limiting]

def fn_join_vals(x):
    return ';'.join([y for y in list(x) if pd.notna(y)])

agg_dict = {'Check Amount':'sum', 
'Contract Purpose': fn_join_vals,
'Budget Code': fn_join_vals,
'Agency': fn_join_vals}

df = df_limited_cols.groupby(temp_cols_for_grouping, as_index=False).agg(agg_dict)
df

FMS ID  Check Amount  \
0         002FINC01   53383178.89   
1       035L22DITFI     276110.00   
2      035L22DRSSCP    1998283.01   
3       035LR20PCPR     307000.00   
4      035LRCA12TEK     154787.43   
...             ...           ...   
16682     998SIMISC    1251580.56   
16683   998SIMTABUS    3500000.00   
16684      998SISEC     250000.00   
16685    998SITRACK   11948020.59   
16686     998TCMISC    3346000.00   

                                        Contract Purpose  \
0                                                          
1      Digital Imaging Equipment;Digital Imaging Equi...   
2                            Dell/EMC Equipment for DRSS   
3      PC, Laptop, Printer Refresh;PC, Laptop, Printe...   
4      Enterprise Storage System;Enterprise Storage S...   
...                                                  ...   
16682  CP66058 SIR ST. GEORGE TRACK AND SIGNAL REPLAC...   
16683  BUS REPLACEMENT;BUS REPLACEMENT ON BEHALF OF N...   
16684     BUS PURCHASE,STATION REPAIRS, SECURITY CAMERAS   
16685  2017 MAINLINE TRACK REPLACEMENT AT VARIOUS LOC...   
16686  CP53717 PURCHASE 61 METROCARD VENDING MACHINES...   

                                             Budget Code  \
0      DUMY (INITIAL COI AUTHORIZATION; FY 2004C);DUM...   
1      TFIE (NYRL, LIBRARY SERVICES CENTER (31-11 THO...   
2        DRS8 (NYPL: SYSTEMWIDE: UPGRADE & EXPAND DIGIT)   
3      LRPC (NYPL, PURCHASE NETWORKED COMPUTER SYSTE)...   
4      EBUS (NYPL SYSTEMWIDE: PURCHASE ENTERPRISE BAK...   
...                                                  ...   
16682  TRT3 (SIRTOA (STATEN ISLAND RAPID TRANSIT OPER...   
16683  RBUS (NYCT: PURCHASE BY MTA OF 745 HYBRD ELECT...   
16684                                                      
16685  SUB1 (MTA /NYCT: CREATE NEW ELECTICAL POWER SU...   
16686  D001 (MTA: PURCAHSE METROCARD FARE COLLECTION)...   

                                                  Agency  
0      Department of Citywide Administrative Services...  
1      New York Research Libraries;New York Research ...  
2                            New York Research Libraries  
3      New York Research Libraries;New York Research ...  
4      New York Research Libraries;New York Research ...  
...                                                  ...  
16682  Transit Authority;TRANSIT AUTHORITY - CAPITAL;...  
16683  TRANSIT AUTHORITY - CAPITAL;TRANSIT AUTHORITY ...  
16684                        TRANSIT AUTHORITY - CAPITAL  
16685  Transit Authority;Transit Authority;Transit Au...  
16686  Transit Authority;Transit Authority;Transit Au...  

[16687 rows x 5 columns]

## Assigning categories based on `Budget Code` and `Contract Purpose`<a class="anchor" id="third-bullet"></a>


In [9]:
# --- create bc_category and cp_category columns
df['bc_category'] = None
df['cp_category'] = None

In [10]:
# --- establish sql db and session
db_url = 'sqlite:///checkbook.db'
engine = create_engine(db_url)
Session = sessionmaker(bind=engine)
session = Session()

table_name = 'capital_projects'
df.to_sql(table_name, engine, if_exists='replace', index=False)

16687

In [11]:
# SQL Query for ITT, Vehicles, and Equipment category

query_itt_vehicles_equipment = """
UPDATE capital_projects SET bc_category = 'ITT, Vehicles, and Equipment'
WHERE (
upper([Budget Code]) LIKE '%VEHICLE%'
OR upper([Budget Code]) LIKE '%AMBULANCE%'
OR upper([Budget Code]) LIKE '%BUSES%'
OR upper([Budget Code]) LIKE '%TRUCK%'
OR upper([Budget Code]) LIKE '%TRAILER%'
OR upper([Budget Code]) LIKE '%VANS%'

OR upper([Budget Code]) LIKE '%EQUIPMENT%'
OR upper([Budget Code]) LIKE '%CRANE%'
OR upper([Budget Code]) LIKE '%FURNITURE%'
OR upper([Budget Code]) LIKE '%PORTABLE%'

OR upper([Budget Code]) LIKE '% IT %'
OR upper([Budget Code]) LIKE '%AUDIO%'
OR upper([Budget Code]) LIKE '%CAMERA%'
OR upper([Budget Code]) LIKE '%COMPUTERS%'
OR upper([Budget Code]) LIKE '%DATA%'
OR upper([Budget Code]) LIKE '%DOITT%'
OR upper([Budget Code]) LIKE '%FISA%'
OR upper([Budget Code]) LIKE '%GIS%'
OR upper([Budget Code]) LIKE '%HARDWARE%'
OR upper([Budget Code]) LIKE '%MAINFRAME%'
OR upper([Budget Code]) LIKE '%MOBILE%'
OR upper([Budget Code]) LIKE '%PHONE%'
OR upper([Budget Code]) LIKE '%PRINTER%'
OR upper([Budget Code]) LIKE '%SERVER%'
OR upper([Budget Code]) LIKE '%SOFTWARE%'
OR upper([Budget Code]) LIKE '%RADIO %'
OR upper([Budget Code]) LIKE '%RADIOS%'
OR upper([Budget Code]) LIKE '%VIDEO%'
OR upper([Budget Code]) LIKE '%VOIP%'
OR upper([Budget Code]) LIKE '%WIFI%'

-- New trigger words
OR upper([Budget Code]) LIKE '%A/V%'
OR upper([Budget Code]) LIKE '% AV %'
OR upper([Budget Code]) LIKE '%ACCESS%UPGRADE%'
OR upper([Budget Code]) LIKE '%AIMS%'
OR upper([Budget Code]) LIKE '%ANALYTICS%'
OR upper([Budget Code]) LIKE '%ANTENA%'
OR upper([Budget Code]) LIKE '%APPARATUS%'
OR upper([Budget Code]) LIKE '%ARREST%PROCESSING%'
OR upper([Budget Code]) LIKE '%ARRESTOR%'
OR upper([Budget Code]) LIKE '%ASSET%TRACKER%'
OR upper([Budget Code]) LIKE '%AUTOMATIC%'
OR upper([Budget Code]) LIKE '%AUTOMATION%'
OR upper([Budget Code]) LIKE '%BARGE%'
OR upper([Budget Code]) LIKE '%BED%'
OR upper([Budget Code]) LIKE '% BMS %'
OR upper([Budget Code]) LIKE '%BOATS%'
OR upper([Budget Code]) LIKE '% BUS %'
OR upper([Budget Code]) LIKE '%BUS'
OR upper([Budget Code]) LIKE '%CART%'
OR upper([Budget Code]) LIKE '%CATHETERIZATION%'
OR upper([Budget Code]) LIKE '%CCTV%'
OR upper([Budget Code]) LIKE '%CHECKBOOK%NYC%'
OR upper([Budget Code]) LIKE '%CISCO%'
OR upper([Budget Code]) LIKE '%CITYTIME%'
OR upper([Budget Code]) LIKE '%COMMUNICATIONS%'
OR upper([Budget Code]) LIKE '%COMMUNICATION%EQUI%'
OR upper([Budget Code]) LIKE '%COMPRESSOR%'
OR upper([Budget Code]) LIKE '%COMPUTER%EQUIP%'
OR upper([Budget Code]) LIKE '%COMPUTER%REPL%'
OR upper([Budget Code]) LIKE '%COMPUTER%SYS%'
OR upper([Budget Code]) LIKE '%COMP%UPGRADE%'
OR upper([Budget Code]) LIKE '%COMPUTERIZED%'
OR upper([Budget Code]) LIKE '%COPIER%'
OR upper([Budget Code]) LIKE '%CT%SCAN%'
OR upper([Budget Code]) LIKE '%CURTAIN%'
OR upper([Budget Code]) LIKE '%CYTOMETER%'
OR upper([Budget Code]) LIKE '%DCTV%'
OR upper([Budget Code]) LIKE '%DESKTOP%'
OR upper([Budget Code]) LIKE '%DEVICE%'
OR upper([Budget Code]) LIKE '%DIGITAL%'
OR upper([Budget Code]) LIKE '%DISINFECTION%'
OR upper([Budget Code]) LIKE '%E-TICKET%'
OR upper([Budget Code]) LIKE '%ECTP%'
OR upper([Budget Code]) LIKE '%ELECTRONIC%'
OR upper([Budget Code]) LIKE '%EMAIL%'
OR upper([Budget Code]) LIKE '%ENCRYPT%'
OR upper([Budget Code]) LIKE '%ENGINE%'
OR upper([Budget Code]) LIKE '%ENTERPRISE%SOLUTIONS%'
OR upper([Budget Code]) LIKE '%EQ%PURCHASE%'
OR upper([Budget Code]) LIKE '%EQUIP%'
OR upper([Budget Code]) LIKE '%FERRY%BO%'
OR upper([Budget Code]) LIKE '%FFE%'
OR upper([Budget Code]) LIKE '%FILM%'
OR upper([Budget Code]) LIKE '%FIREBOAT%'
OR upper([Budget Code]) LIKE '%FLAT%BED%'
OR upper([Budget Code]) LIKE '%FLEET%'
OR upper([Budget Code]) LIKE '%FMS%'
OR upper([Budget Code]) LIKE '%FORKLIFT%'
OR upper([Budget Code]) LIKE '%GENERATOR%'
OR upper([Budget Code]) LIKE '%GPS%'
OR upper([Budget Code]) LIKE '%HDTV%'
OR upper([Budget Code]) LIKE '%HELICOPTER%'
OR upper([Budget Code]) LIKE '%HHS%ACC%'
OR upper([Budget Code]) LIKE '%INFO%MANAGE%'
OR upper([Budget Code]) LIKE '%INFO%SECURITY%'
OR upper([Budget Code]) LIKE '%INTERFACE%'
OR upper([Budget Code]) LIKE 'IT %'
OR upper([Budget Code]) LIKE '% IT'
OR upper([Budget Code]) LIKE '%IVR %'
OR upper([Budget Code]) LIKE '%KEYWORD%'
OR upper([Budget Code]) LIKE '%KITS%'
OR upper([Budget Code]) LIKE '%LADDER%'
OR upper([Budget Code]) LIKE '% LAN %'
OR upper([Budget Code]) LIKE '%LUNG%UNIT%'
OR upper([Budget Code]) LIKE '%MACHINE%'
OR upper([Budget Code]) LIKE '%MAMOGRAM%'
OR upper([Budget Code]) LIKE '%MANHOLE%COVER%'
OR upper([Budget Code]) LIKE '%MANHOLE%RING%'
OR upper([Budget Code]) LIKE '%MANOMETRY%'
OR upper([Budget Code]) LIKE '%MED%EQMT%'
OR upper([Budget Code]) LIKE '%MEDIA%UPGRADE%'
OR upper([Budget Code]) LIKE '%MICROFILM%'
OR upper([Budget Code]) LIKE '%MICROSCOPE%'
OR upper([Budget Code]) LIKE '%MICROSPECT%'
OR upper([Budget Code]) LIKE '%MINI%BUS%'
OR upper([Budget Code]) LIKE '%MINI%VAN%'
OR upper([Budget Code]) LIKE '%MOBILIZING%'
OR upper([Budget Code]) LIKE '%MONITOR%'
OR upper([Budget Code]) LIKE '%MOORING%UPGRADE%'
OR upper([Budget Code]) LIKE '%MOSAICS%'
OR upper([Budget Code]) LIKE '%MOVEABLE%'
OR upper([Budget Code]) LIKE '%MRI%'
OR upper([Budget Code]) LIKE '%NETWORK%'
OR upper([Budget Code]) LIKE '%NOVAS%'
OR upper([Budget Code]) LIKE '%NYCAPP%'
OR upper([Budget Code]) LIKE '%NYCAPS%'
OR upper([Budget Code]) LIKE '%NYCSERV%'
OR upper([Budget Code]) LIKE '%OAISIS%'
OR upper([Budget Code]) LIKE '%PAPERLESS%'
OR upper([Budget Code]) LIKE '%PASSENGER%VAN%'
OR upper([Budget Code]) LIKE '%PIANO%'
OR upper([Budget Code]) LIKE '%PORTAL%'
OR upper([Budget Code]) LIKE '%POWER%WASHER%'
OR upper([Budget Code]) LIKE '%PRINTING%'
OR upper([Budget Code]) LIKE '%PROJECTORS%'
OR upper([Budget Code]) LIKE '%PROJECTION%SYS%'
OR upper([Budget Code]) LIKE '%PROJECTION%SOUND%SYSTEM%'
OR upper([Budget Code]) LIKE '%PUMPERS%'
OR upper([Budget Code]) LIKE '%PURCHASE%'
OR upper([Budget Code]) LIKE '%REAL%TIME%INFO%'
OR upper([Budget Code]) LIKE '%REAL%TIME%SIGN%'
OR upper([Budget Code]) LIKE '%RECORDER%'
OR upper([Budget Code]) LIKE '%REFRIGERATOR%'
OR upper([Budget Code]) LIKE '%REPORTING%'
OR upper([Budget Code]) LIKE '%RESPONSE%BOAT%'
OR upper([Budget Code]) LIKE '% RIG %'
OR upper([Budget Code]) LIKE '%ROBOT%'
OR upper([Budget Code]) LIKE '%ROUTER%'
OR upper([Budget Code]) LIKE '%SATELLITE%'
OR upper([Budget Code]) LIKE '%SCANNER%'
OR upper([Budget Code]) LIKE '%SCANNING%'
OR upper([Budget Code]) LIKE '%SCHOOL%BUS%'
OR upper([Budget Code]) LIKE '%SELF%CHECK%'
OR upper([Budget Code]) LIKE '%SEQUENCER%'
OR upper([Budget Code]) LIKE '%SHUTTLE%'
OR upper([Budget Code]) LIKE '%SIMULATOR%'
OR upper([Budget Code]) LIKE '%SKID%STEER%'
OR upper([Budget Code]) LIKE '%SPECTROMETER%'
OR upper([Budget Code]) LIKE '%SPECTROPHMETER%'
OR upper([Budget Code]) LIKE '%SUBURBANS%'
OR upper([Budget Code]) LIKE '%SURVEILLANCE%'
OR upper([Budget Code]) LIKE '%SONOGRAM%'
OR upper([Budget Code]) LIKE '%SOUND%SYST%'
OR upper([Budget Code]) LIKE '%STREAMING%'
OR upper([Budget Code]) LIKE '%SWEEPER%'
OR upper([Budget Code]) LIKE '%SYSTEM%INTEGRATOR%'
OR upper([Budget Code]) LIKE '%SYSTEM%UPGRADE%'
OR upper([Budget Code]) LIKE '% TABLES%'
OR upper([Budget Code]) LIKE '% TECH%'
OR upper([Budget Code]) LIKE 'TECH%'
OR upper([Budget Code]) LIKE '%TELECOM%'
OR upper([Budget Code]) LIKE '%TELEMETRY%'
OR upper([Budget Code]) LIKE '%TOOLCAT%'
OR upper([Budget Code]) LIKE '%TRACTOR%'
OR upper([Budget Code]) LIKE '%TRAM%'
OR upper([Budget Code]) LIKE '%TRANSPORTATION%BUS%'
OR upper([Budget Code]) LIKE '%TROLLEY%'
OR upper([Budget Code]) LIKE '%ULTRASOUND%'
OR upper([Budget Code]) LIKE '%VAN'
OR upper([Budget Code]) LIKE '%VASCULAR%'
OR upper([Budget Code]) LIKE '%VENDING%'
OR upper([Budget Code]) LIKE '%VESSEL%'
OR upper([Budget Code]) LIKE '%VIRTUAL%'
OR upper([Budget Code]) LIKE '%VOICE%'
OR upper([Budget Code]) LIKE '%WAGON%'
OR upper([Budget Code]) LIKE '%WEB%PROXY%'
OR upper([Budget Code]) LIKE '%WEBSITE%'
OR upper([Budget Code]) LIKE '%WI-FI%'
OR upper([Budget Code]) LIKE '%WIRELESS%'
OR upper([Budget Code]) LIKE '%WIRETAP%'
OR upper([Budget Code]) LIKE '%WORK%STATION%'
OR upper([Budget Code]) LIKE '% X%RAY%'
OR upper([Budget Code]) LIKE '%ZAMBONI%'

-- Systems
OR upper([Budget Code]) LIKE '%ADDRESS%SYST%'
OR upper([Budget Code]) LIKE '%ADMIN%SYS%'
OR upper([Budget Code]) LIKE '%ASSET%SYS%'
OR upper([Budget Code]) LIKE '%ASSIGNMENT%SYST%'
OR upper([Budget Code]) LIKE '%ALARM%SYST%'
OR upper([Budget Code]) LIKE '%AUTOMATED%SYST%'
OR upper([Budget Code]) LIKE '%BILLING%SYST%'
OR upper([Budget Code]) LIKE '%BIOVAULT%SYST%'
OR upper([Budget Code]) LIKE '%COMMUNICATION%SYST%'
OR upper([Budget Code]) LIKE '%CRIME%SYST%'
OR upper([Budget Code]) LIKE '%ENROLLMENT%SYS%'
OR upper([Budget Code]) LIKE '%EQ %SYS%'
OR upper([Budget Code]) LIKE '%FAST%PASS%SYS%'
OR upper([Budget Code]) LIKE '%FILING%SYS%'
OR upper([Budget Code]) LIKE '%FIN% SYS%'
OR upper([Budget Code]) LIKE '%IMAGING%SYST%'
OR upper([Budget Code]) LIKE '%INFO%SYST%'
OR upper([Budget Code]) LIKE '%INGEST%SYST%'
OR upper([Budget Code]) LIKE '%INTELL%SYST%'
OR upper([Budget Code]) LIKE '%LAUNDRY%SYS%'
OR upper([Budget Code]) LIKE '%LICENSING%SYS%'
OR upper([Budget Code]) LIKE '%LOAD%SYS%'
OR upper([Budget Code]) LIKE '%MANAGEMENT%SYST%'
OR upper([Budget Code]) LIKE '%MEASUREMENT%SYST%'
OR upper([Budget Code]) LIKE '%MED%SYST%'
OR upper([Budget Code]) LIKE '%MEDIA%SYST%'
OR upper([Budget Code]) LIKE '%PRESERVATION%SYST%'
OR upper([Budget Code]) LIKE '%PREVENTION%SYST%'
OR upper([Budget Code]) LIKE '%PROCESSING%SYST%'
OR upper([Budget Code]) LIKE '%SECURITY%SYST%'
OR upper([Budget Code]) LIKE '%SYSTEMWIDE%SYST%'
OR upper([Budget Code]) LIKE '%TAX%SYST%'
OR upper([Budget Code]) LIKE '%TICKET%SYST%'
OR upper([Budget Code]) LIKE '%TRACKING%SYST%'
OR upper([Budget Code]) LIKE '%WORKFORCE%SYST%'
)
AND( upper([Budget Code]) NOT LIKE '%GARAGE%' );
"""

In [12]:
# SQL Query for Lump Sum category
query_lump_sum = """
UPDATE capital_projects SET bc_category = 'Lump Sum'
WHERE (
upper([Budget Code]) LIKE '%LUMP SUM%'
OR upper([Budget Code]) LIKE '% FUND%'
OR upper([Budget Code]) LIKE 'FUND%'
OR upper([Budget Code]) LIKE '%SURVEY%'
OR upper([Budget Code]) LIKE '%SUPERVISION%'
OR upper([Budget Code]) LIKE '%PROGRAM%'

--new
OR upper([Budget Code]) LIKE '%10%YEAR%PLAN%'
OR upper([Budget Code]) LIKE '%ACQUISITION%CITYWIDE%'
OR upper([Budget Code]) LIKE '%AGENCY%PROPOSED%PROJECT%'
OR upper([Budget Code]) LIKE '%AGREEMENT%'
OR upper([Budget Code]) LIKE '%APPLICATION%'
OR upper([Budget Code]) LIKE '%ASSESSMENT%'
OR upper([Budget Code]) LIKE '%ASSOC%'
OR upper([Budget Code]) LIKE '%AUDITOR%'
OR upper([Budget Code]) LIKE '%AVIATION%'
OR upper([Budget Code]) LIKE '%BIOSWALES%STORMWATER%'
OR upper([Budget Code]) LIKE '%BPL%INFRASTRUCTURE%RECONSTRUCTION%'
OR upper([Budget Code]) LIKE '%BUNDLE%PROJECTS%'
OR upper([Budget Code]) LIKE '%CAPITAL%COMP%'
OR upper([Budget Code]) LIKE '%CAMPAIGN%'
OR upper([Budget Code]) LIKE '%CALTHOLIC%MANAGE%'
OR upper([Budget Code]) LIKE '%CITY%WIDE%ACQUISITION%'
OR upper([Budget Code]) LIKE '%CITY%WIDE%MANAGEMENT%'
OR upper([Budget Code]) LIKE '%CITY%WIDE%MEASURES%'
OR upper([Budget Code]) LIKE '%CITY%WIDE%SECURITY%'
OR upper([Budget Code]) LIKE '%COACH%'
OR upper([Budget Code]) LIKE '%COALITION%'
OR upper([Budget Code]) LIKE '%COMB%'
OR upper([Budget Code]) LIKE '%COMMITTEE%'
OR upper([Budget Code]) LIKE '%COMPLIANCE%'
OR upper([Budget Code]) LIKE '%COMPUTER%PROG%'
OR upper([Budget Code]) LIKE '%CONTRACT%'
OR upper([Budget Code]) LIKE '%CONSERVATION%'
OR upper([Budget Code]) LIKE '%CONSULTANT%'
OR upper([Budget Code]) LIKE '%CONTRACTS%'
OR upper([Budget Code]) LIKE '%COOKING%PROJECT%'
OR upper([Budget Code]) LIKE '%COOPERATIVE%'
OR upper([Budget Code]) LIKE '%CORP%'
OR upper([Budget Code]) LIKE '%COUNCIL%'
OR upper([Budget Code]) LIKE '%DISTRIBUTION%FY%'
OR upper([Budget Code]) LIKE '%ELLA%FY%'
OR upper([Budget Code]) LIKE '%EMERGENCY%CONT%'
OR upper([Budget Code]) LIKE '%EMERGENCY%SAFETY%SYSTEMS%'
OR upper([Budget Code]) LIKE '%EXAMINATION%'
OR upper([Budget Code]) LIKE '%FERRY%PROJECTS%'
OR upper([Budget Code]) LIKE '%FUND'
OR upper([Budget Code]) LIKE '%GENERAL%'
OR upper([Budget Code]) LIKE '%HUD %'
OR upper([Budget Code]) LIKE '%HOLD %'
OR upper([Budget Code]) LIKE '% HOLD%'
OR upper([Budget Code]) LIKE '%IFA %'
OR upper([Budget Code]) LIKE '%IMPACT%STATEMENT%'
OR upper([Budget Code]) LIKE '%IMPROVEMENT%CITYWIDE%'
OR upper([Budget Code]) LIKE '% INC%'
OR upper([Budget Code]) LIKE '%INITIATIVE%'
OR upper([Budget Code]) LIKE '%INTITIATIVE%'
OR upper([Budget Code]) LIKE '%INSPCTN%'
OR upper([Budget Code]) LIKE '%INSPECTN%'
OR upper([Budget Code]) LIKE '%INSPECTION%'
OR upper([Budget Code]) LIKE '%INVESTIG%'
OR upper([Budget Code]) LIKE '%JOB%ORDER%CONTRACT%'
OR upper([Budget Code]) LIKE '% LLC%'
OR upper([Budget Code]) LIKE '%LUMP%'
OR upper([Budget Code]) LIKE '%MANAGEMENT%'
OR upper([Budget Code]) LIKE '%MAINTENANCE%'
OR upper([Budget Code]) LIKE '%MASTER%PLAN%'
OR upper([Budget Code]) LIKE '%MGNT%SVCS%'
OR upper([Budget Code]) LIKE '%MGNT%SVCES%'
OR upper([Budget Code]) LIKE '%MISC %'
OR upper([Budget Code]) LIKE '%MISC. %'
OR upper([Budget Code]) LIKE '%MISCELANNOUES%'
OR upper([Budget Code]) LIKE '%MITIGATION%PGM%'
OR upper([Budget Code]) LIKE '%MITIGATION%PROGRAM%'
OR upper([Budget Code]) LIKE '%MONITORING%'
OR upper([Budget Code]) LIKE '%NATIONAL%ASSOC%'
OR upper([Budget Code]) LIKE '%NEW%NEED%'
OR upper([Budget Code]) LIKE '%OPPORTUNIT%'
OR upper([Budget Code]) LIKE '%ORGANIZATION%'
OR upper([Budget Code]) LIKE '%OVERSIGHT%'
OR upper([Budget Code]) LIKE '%PARTNERSHIP%'
OR upper([Budget Code]) LIKE '%PED%SAFETY%'
OR upper([Budget Code]) LIKE '%PEDESTRIAN%SAFETY%'
OR upper([Budget Code]) LIKE '%PEDESTRIAN%SYS%'
OR upper([Budget Code]) LIKE '%PILOT%'
OR upper([Budget Code]) LIKE '%PLANNED%PARENTHOOD%OF%NYC'
OR upper([Budget Code]) LIKE '%PRIORITY%GRID%'
OR upper([Budget Code]) LIKE '%PROCUREMENT%'
OR upper([Budget Code]) LIKE '%PROFESSIONAL%SERVICE%'
OR upper([Budget Code]) LIKE '%PROJECT%RENEWAL'
OR upper([Budget Code]) LIKE '%PROG%'
OR upper([Budget Code]) LIKE '%PROTECTION%CITYWIDE%'
OR upper([Budget Code]) LIKE '%REIMBURSEMENT%'
OR upper([Budget Code]) LIKE '%RESILIENCY%MEASURES%'
OR upper([Budget Code]) LIKE '%RESOURCES%'
OR upper([Budget Code]) LIKE '%REVITAL%PROJECT%'
OR upper([Budget Code]) LIKE '%SAFE%COMMUNIT%'
OR upper([Budget Code]) LIKE '%SAFE%ROUTE%SCHOOL%'
OR upper([Budget Code]) LIKE '%SAFE%ROUTE%TO %'
OR upper([Budget Code]) LIKE '%SAFETY%IMPROVEMENT%'
OR upper([Budget Code]) LIKE '%SAFETY%PROJECT%'
OR upper([Budget Code]) LIKE '%SAMPLING%'
OR upper([Budget Code]) LIKE 'SCA%'
OR upper([Budget Code]) LIKE '%SE%&%WM%'
OR upper([Budget Code]) LIKE '%SEEDING%'
OR upper([Budget Code]) LIKE '%SERVICE%'
OR upper([Budget Code]) LIKE '%SERVICE%AGREEMENT%'
OR upper([Budget Code]) LIKE '%SEWER%WATER%'
OR upper([Budget Code]) LIKE '%SEWER%WM%'
OR upper([Budget Code]) LIKE '%SOCIETY%'
OR upper([Budget Code]) LIKE '%STAFF%'
OR upper([Budget Code]) LIKE '%STORM%WM%'
OR upper([Budget Code]) LIKE '%STORM%WATER%MANAGEMENT%'
OR upper([Budget Code]) LIKE '%STORM%WATER%MGMT%'
OR upper([Budget Code]) LIKE '%STRATEGIC%PLAN%'
OR upper([Budget Code]) LIKE '%STUDIES%'
OR upper([Budget Code]) LIKE '%SUPPORT%'
OR upper([Budget Code]) LIKE '%SURPLUS%'
OR upper([Budget Code]) LIKE '%SWER%WM%'
OR upper([Budget Code]) LIKE '%SWR%WM%'
OR upper([Budget Code]) LIKE '%SWR%W/M%'
OR upper([Budget Code]) LIKE '%SYSTEMWIDE%RENO%NEW%EXPAN%'
OR upper([Budget Code]) LIKE '%TASK%ORDER%'
OR upper([Budget Code]) LIKE '%TEN%YEAR%PLAN%'
OR upper([Budget Code]) LIKE '%TESTING%'
OR upper([Budget Code]) LIKE '%TORTS%'
OR upper([Budget Code]) LIKE '%TREES%'	
OR upper([Budget Code]) LIKE '%ULURP%'
OR upper([Budget Code]) LIKE '%VARIOUS%LOCATIONS%'
OR upper([Budget Code]) LIKE '%VARIOUS%WORK%'
OR upper([Budget Code]) LIKE '%VISION%ZERO%'
OR upper([Budget Code]) LIKE '%VOLUNTEERS%'
OR upper([Budget Code]) LIKE '%W/M%SEWER%'
OR upper([Budget Code]) LIKE '%W/M%SWR%'
OR upper([Budget Code]) LIKE '%WATER%MAIN%SEWER%'
OR upper([Budget Code]) LIKE '%WM%SEWER%'
OR upper([Budget Code]) LIKE '%WM%SWR%'

)
AND( upper([Budget Code]) NOT LIKE '%SPACE%')
AND( upper([Budget Code]) NOT LIKE '%RESTOR%' )
AND bc_category IS NULL;
"""

In [13]:
# SQL Query for Fixed Asset categorization

query_fixed_asset = """
UPDATE capital_projects SET bc_category = 'Fixed Asset'
WHERE (
upper([Budget Code]) LIKE '%AUDITORIUM%'
OR upper([Budget Code]) LIKE '%BASIN%'
OR upper([Budget Code]) LIKE '%BATHROOM%'
OR upper([Budget Code]) LIKE '%BOILER%'
OR upper([Budget Code]) LIKE '%BORING%'
OR upper([Budget Code]) LIKE '%BLVD%'
OR upper([Budget Code]) LIKE '%BRIDGE%'
OR upper([Budget Code]) LIKE '%BUILDING%'
OR upper([Budget Code]) LIKE '%BULKHEAD%'
OR upper([Budget Code]) LIKE '%CAFETERIA%'
OR upper([Budget Code]) LIKE '%CANAL%'
OR upper([Budget Code]) LIKE '%CEILING%'
OR upper([Budget Code]) LIKE '%CENTER%'
OR upper([Budget Code]) LIKE '%CLIMATE%'
OR upper([Budget Code]) LIKE '%CONSTRUCTION%'
OR upper([Budget Code]) LIKE '%COOLING%'
OR upper([Budget Code]) LIKE '%CULVERT%'
OR upper([Budget Code]) LIKE '%DAY CARE%'
OR upper([Budget Code]) LIKE '%DEPOT%'
OR upper([Budget Code]) LIKE '%ELECTRICAL%'
OR upper([Budget Code]) LIKE '%ELEVATOR%'
OR upper([Budget Code]) LIKE '%ESCALATOR%'
OR upper([Budget Code]) LIKE '%EXTERIOR%'
OR upper([Budget Code]) LIKE '%FACILITY%'
OR upper([Budget Code]) LIKE '%FENC%'
OR upper([Budget Code]) LIKE '%FIELD%'
OR upper([Budget Code]) LIKE '%FIXTURE%'
OR upper([Budget Code]) LIKE '%FLOOR%'
OR upper([Budget Code]) LIKE '%GARAGE%'
OR upper([Budget Code]) LIKE '%GARDEN%'
OR upper([Budget Code]) LIKE '%GREENWAY%'
OR upper([Budget Code]) LIKE '%GYM%'
OR upper([Budget Code]) LIKE '%HALL%'
OR upper([Budget Code]) LIKE '%HEATING%'
OR upper([Budget Code]) LIKE '%HOSPITAL%'
OR upper([Budget Code]) LIKE '%HOUSE%'
OR upper([Budget Code]) LIKE '%HVAC%'
OR upper([Budget Code]) LIKE '%INTERIOR%'
OR upper([Budget Code]) LIKE '%KITCHEN%'
OR upper([Budget Code]) LIKE '%LAB%'
OR upper([Budget Code]) LIKE '%LANDING%'
OR upper([Budget Code]) LIKE '%LIBRARY%'
OR upper([Budget Code]) LIKE '%LIGHTING%'
OR upper([Budget Code]) LIKE '%MASONRY%'
OR upper([Budget Code]) LIKE '%MILLING%'
OR upper([Budget Code]) LIKE '%MTS%'
OR upper([Budget Code]) LIKE '%MUSEUM%'
OR upper([Budget Code]) LIKE '%PARAPET%'
OR upper([Budget Code]) LIKE '%PARK%'
OR upper([Budget Code]) LIKE '%PIER%'
OR upper([Budget Code]) LIKE '%PIPE%'
OR upper([Budget Code]) LIKE '%PIPING%'
OR upper([Budget Code]) LIKE '%PLANT%'
OR upper([Budget Code]) LIKE '%PLAYGROUND%'
OR upper([Budget Code]) LIKE '%PLAZA%'
OR upper([Budget Code]) LIKE '%POOL%'
OR upper([Budget Code]) LIKE '%RAMP%'
OR upper([Budget Code]) LIKE '%RECON%'
OR upper([Budget Code]) LIKE '%REHAB%'
OR upper([Budget Code]) LIKE '%RENOVAT%'
OR upper([Budget Code]) LIKE '%REPLACE%'
OR upper([Budget Code]) LIKE '%RESTORATION%'
OR upper([Budget Code]) LIKE '%ROOF%'
OR upper([Budget Code]) LIKE '%ROOM%'
OR upper([Budget Code]) LIKE '%SEWER%'
OR upper([Budget Code]) LIKE '%SHELTER%'
OR upper([Budget Code]) LIKE '%SIDEWALK%'
OR upper([Budget Code]) LIKE '%SITE ACQ%'
OR upper([Budget Code]) LIKE '%SPRAY%BOOTH%'
OR upper([Budget Code]) LIKE '%STADIUM%'
OR upper([Budget Code]) LIKE '%STATION%'
OR upper([Budget Code]) LIKE '%STREET%'
OR upper([Budget Code]) LIKE '%TERMINAL%'
OR upper([Budget Code]) LIKE '%THEATER%'
OR upper([Budget Code]) LIKE '%TOILET%'
OR upper([Budget Code]) LIKE '%TUNNEL%'
OR upper([Budget Code]) LIKE '%VENTILATION%'
OR upper([Budget Code]) LIKE '%WASTE%'
OR upper([Budget Code]) LIKE '%WINDOW%'
OR upper([Budget Code]) LIKE '%WPC%'
OR upper([Budget Code]) LIKE '%WWTP%'

--new
OR upper([Budget Code]) LIKE '%A/C%'
OR upper([Budget Code]) LIKE '%ACQ%'
OR upper([Budget Code]) LIKE '%ACQUISITION%'
OR upper([Budget Code]) LIKE '%AERATION%'
OR upper([Budget Code]) LIKE '%ADA%'
OR upper([Budget Code]) LIKE '%ADDITION%'
OR upper([Budget Code]) LIKE '%AIR%BOX%'
OR upper([Budget Code]) LIKE '%AIR%CON%'
OR upper([Budget Code]) LIKE '%ALARM%BOX%'
OR upper([Budget Code]) LIKE '%ANNEX%'
OR upper([Budget Code]) LIKE '%ARENA%'
OR upper([Budget Code]) LIKE '%ART%'
OR upper([Budget Code]) LIKE '%ASBESTOS%'
OR upper([Budget Code]) LIKE '%ASPHALT%'
OR upper([Budget Code]) LIKE '%ATRIUM%'
OR upper([Budget Code]) LIKE '%ATC%'
OR upper([Budget Code]) LIKE '%AQUARIUM%'
OR upper([Budget Code]) LIKE '%BACKFLOW%SYS%'
OR upper([Budget Code]) LIKE '%BARRIER%'
OR upper([Budget Code]) LIKE '%BOAT%LAUNCH%'
OR upper([Budget Code]) LIKE '%BIKEWAY%'
OR upper([Budget Code]) LIKE '%BIN%'
OR upper([Budget Code]) LIKE '%BIOSWALE%'
OR upper([Budget Code]) LIKE '%BLDG%'
OR upper([Budget Code]) LIKE '%BLDG%ACQ%'
OR upper([Budget Code]) LIKE '%BLUEBELT%'
OR upper([Budget Code]) LIKE '%BOARDWALK%'
OR upper([Budget Code]) LIKE '%BURNER%'
OR upper([Budget Code]) LIKE '%CABLING%'
OR upper([Budget Code]) LIKE '%CALL%BOX%'
OR upper([Budget Code]) LIKE '%CAMPUS%'
OR upper([Budget Code]) LIKE '%CAR%PORT%'
OR upper([Budget Code]) LIKE '%CATWALK%'
OR upper([Budget Code]) LIKE '%CEMETERY%'
OR upper([Budget Code]) LIKE '%CHAMBER%'
OR upper([Budget Code]) LIKE '%CHILLER%'
OR upper([Budget Code]) LIKE '%CLINIC%'
OR upper([Budget Code]) LIKE '%COMFORT%ST%'
OR upper([Budget Code]) LIKE '%COMPACTOR%'
OR upper([Budget Code]) LIKE '%COMPUTER%LAB%'
OR upper([Budget Code]) LIKE '%COMPUTER%ROOM%'
OR upper([Budget Code]) LIKE '%CONDUIT%'
OR upper([Budget Code]) LIKE '%CONSOLIDATION%'
OR upper([Budget Code]) LIKE '%CONSTRUCT%'
OR upper([Budget Code]) LIKE '%CONVERSION%'
OR upper([Budget Code]) LIKE '%CORRIDOR%'
OR upper([Budget Code]) LIKE '%COURT%'
OR upper([Budget Code]) LIKE '% CTS %'
OR upper([Budget Code]) LIKE 'CTS %'
OR upper([Budget Code]) LIKE '% CTS'
OR upper([Budget Code]) LIKE '%CSO%'
OR upper([Budget Code]) LIKE '%DAM%'
OR upper([Budget Code]) LIKE '%DEMOLITION%'
OR upper([Budget Code]) LIKE '%DEVELOPMENT%'
OR upper([Budget Code]) LIKE '%DIGESTER%'
OR upper([Budget Code]) LIKE '%DIMMER%SYST%'
OR upper([Budget Code]) LIKE '%DOCK%'
OR upper([Budget Code]) LIKE '%DOG%RUN%'
OR upper([Budget Code]) LIKE '%DOOR%'
OR upper([Budget Code]) LIKE '%DORM%'
OR upper([Budget Code]) LIKE '%DUCT%'
OR upper([Budget Code]) LIKE '%EE%UPGRADE%'
OR upper([Budget Code]) LIKE '%EJECTOR%'
OR upper([Budget Code]) LIKE '%ELECTRIC%SYS%'
OR upper([Budget Code]) LIKE '%ELEC%UPGRADE%'
OR upper([Budget Code]) LIKE '%ENCLOS%'
OR upper([Budget Code]) LIKE '%ENERGY%UPGRADE%'
OR upper([Budget Code]) LIKE '%ENTRANCE%'
OR upper([Budget Code]) LIKE '%ENTRY'
OR upper([Budget Code]) LIKE '%ESPLANADE%'
OR upper([Budget Code]) LIKE '%EXHAUST%SYSTEM%'
OR upper([Budget Code]) LIKE '%EXHIBIT%'
OR upper([Budget Code]) LIKE '%EXPANSION%'
OR upper([Budget Code]) LIKE '%EXTENSION%'
OR upper([Budget Code]) LIKE '%FACADE%'
OR upper([Budget Code]) LIKE '%FACILIT%'
OR upper([Budget Code]) LIKE '%FIRE%ALAR%'
OR upper([Budget Code]) LIKE '%FIRE%SAFETY%'
OR upper([Budget Code]) LIKE '%FIRE%SYSTEM%'
OR upper([Budget Code]) LIKE '%FIRE%UPGRADE%'
OR upper([Budget Code]) LIKE '%FLAG%'
OR upper([Budget Code]) LIKE '%FOUNTAIN%'
OR upper([Budget Code]) LIKE '%GALLERY%'
OR upper([Budget Code]) LIKE '%GAS%'
OR upper([Budget Code]) LIKE '%GATE%'
OR upper([Budget Code]) LIKE '%GOLF%COURSE%'
OR upper([Budget Code]) LIKE '%GOVERNORS%ISLAND%'
OR upper([Budget Code]) LIKE '% GRAT%'
OR upper([Budget Code]) LIKE '%GUARDS%'
OR upper([Budget Code]) LIKE '%HEAT%'
OR upper([Budget Code]) LIKE '%HIGH%LINE%'
OR upper([Budget Code]) LIKE '%HOT%WATER%'
OR upper([Budget Code]) LIKE '%INFRASTRUCTURE%'
OR upper([Budget Code]) LIKE '%INSTALLATION%'
OR upper([Budget Code]) LIKE '%INTERCONNECTION%'
OR upper([Budget Code]) LIKE '%IRRIGATION%'
OR upper([Budget Code]) LIKE '%LAMP%'
OR upper([Budget Code]) LIKE '%LAWNS%'
OR upper([Budget Code]) LIKE '%LEASE%RENEW%'
OR upper([Budget Code]) LIKE '%LIGHTS%'
OR upper([Budget Code]) LIKE '%LIGHTPOLE%'
OR upper([Budget Code]) LIKE '%LOCAL%LAW%11%'
OR upper([Budget Code]) LIKE '%LOUNGE%'
OR upper([Budget Code]) LIKE '%MARINA%'
OR upper([Budget Code]) LIKE '%MEDIAN%'
OR upper([Budget Code]) LIKE '%MICROFILTRATION%'
OR upper([Budget Code]) LIKE '%MODERNIZATION%'
OR upper([Budget Code]) LIKE '% MOVE %'
OR upper([Budget Code]) LIKE '%NEW%BRANCH%'
OR upper([Budget Code]) LIKE 'NDF%'
OR upper([Budget Code]) LIKE '%OBSERVATORY%'
OR upper([Budget Code]) LIKE '%OFFICES%'
OR upper([Budget Code]) LIKE '%OUTDOOR%SPACE%'
OR upper([Budget Code]) LIKE '%OUTFITTIN%'
OR upper([Budget Code]) LIKE '%OUT-FITTIN%'
OR upper([Budget Code]) LIKE '%OVERPASS%'
OR upper([Budget Code]) LIKE '%PASSIVE%LANDSCAPE%'
OR upper([Budget Code]) LIKE '%PATH%'
OR upper([Budget Code]) LIKE '%PAVEMENT%'
OR upper([Budget Code]) LIKE '% PEN %'
OR upper([Budget Code]) LIKE '%PEDESTALS%'
OR upper([Budget Code]) LIKE '%PLDG%'
OR upper([Budget Code]) LIKE '%PLGD%'
OR upper([Budget Code]) LIKE '%PLGRD%'
OR upper([Budget Code]) LIKE '%PLYG%'
OR upper([Budget Code]) LIKE '%PRINTSHOP%'
OR upper([Budget Code]) LIKE '%PROMENADE%'
OR upper([Budget Code]) LIKE '%PUMP%'
OR upper([Budget Code]) LIKE '%PV PANELS%'
OR upper([Budget Code]) LIKE '%REFURB%'
OR upper([Budget Code]) LIKE '%RELOCAT%'
OR upper([Budget Code]) LIKE '%REMOVAL%'
OR upper([Budget Code]) LIKE '%RENO%'
OR upper([Budget Code]) LIKE '%REPAIR%SHOP%'
OR upper([Budget Code]) LIKE '%REP% WM %'
OR upper([Budget Code]) LIKE '%RESTITUTION%'
OR upper([Budget Code]) LIKE '%RESURFACING%'
OR upper([Budget Code]) LIKE '%RETAINING%WALL%'
OR upper([Budget Code]) LIKE '%RETRO%FIT%'
OR upper([Budget Code]) LIKE '%RIDING%RING%'
OR upper([Budget Code]) LIKE '%RINK%'
OR upper([Budget Code]) LIKE '%RD%IMPROVEMENT%'
OR upper([Budget Code]) LIKE '%ROAD%IMPROVEMENT%'
OR upper([Budget Code]) LIKE '%ROAD%WIDE%'
OR upper([Budget Code]) LIKE '%ROPE%'
OR upper([Budget Code]) LIKE '%SCHOOLYARDS%'
OR upper([Budget Code]) LIKE '%SBS %'
OR upper([Budget Code]) LIKE '% SBS%'
OR upper([Budget Code]) LIKE '%SEAT %'
OR upper([Budget Code]) LIKE '%SEATING%'
OR upper([Budget Code]) LIKE '%SECURITY%INSTALL%'
OR upper([Budget Code]) LIKE '%SECURITY%SYS%'
OR upper([Budget Code]) LIKE '%SECURITY%UPG%'
OR upper([Budget Code]) LIKE '%SHAFT%'
OR upper([Budget Code]) LIKE '%SHED%'
OR upper([Budget Code]) LIKE '%SHORELINE%'
OR upper([Budget Code]) LIKE '%SHOWER%'
OR upper([Budget Code]) LIKE '%SIGN%'
OR upper([Budget Code]) LIKE '%SKATE P%'
OR upper([Budget Code]) LIKE '%SKYLIGHT%'
OR upper([Budget Code]) LIKE '%SMOKE%SYSTEM%'
OR upper([Budget Code]) LIKE '%SOLAR%PHOTO%'
OR upper([Budget Code]) LIKE '%SOLAR%PV%'
OR upper([Budget Code]) LIKE '%SPACE%'
OR upper([Budget Code]) LIKE '%SOCCER%'
OR upper([Budget Code]) LIKE '%SPORT%AREA%'
OR upper([Budget Code]) LIKE '%SPORT%COMPLEX%'
OR upper([Budget Code]) LIKE '%SPACE%IMPRO%'
OR upper([Budget Code]) LIKE '%SPRINKLER%'
OR upper([Budget Code]) LIKE '%STABILIZATION%'
OR upper([Budget Code]) LIKE '%STAGE%'
OR upper([Budget Code]) LIKE '%STAIR%'
OR upper([Budget Code]) LIKE '%STAIRCASE%'
OR upper([Budget Code]) LIKE '%STEAM%'
OR upper([Budget Code]) LIKE '%STM%SWR%'
OR upper([Budget Code]) LIKE '%STORM%SWR%'
OR upper([Budget Code]) LIKE '%STORAGE%'
OR upper([Budget Code]) LIKE '%STRM%SWR%'
OR upper([Budget Code]) LIKE '%STRUCTURE%'
OR upper([Budget Code]) LIKE '%STUDIO%'
OR upper([Budget Code]) LIKE '%SWING%'
OR upper([Budget Code]) LIKE '%SWR%'
OR upper([Budget Code]) LIKE '%TANK%'
OR upper([Budget Code]) LIKE '%TOW%POUND%'
OR upper([Budget Code]) LIKE '%TOWER%'
OR upper([Budget Code]) LIKE '%TRAIL%'
OR upper([Budget Code]) LIKE '%TRNG%RM%'
OR upper([Budget Code]) LIKE '%TRUNK%'
OR upper([Budget Code]) LIKE '%TURNSTILE%'
OR upper([Budget Code]) LIKE '%UNDERGROUND%'
OR upper([Budget Code]) LIKE '%UNDERPASS%'
OR upper([Budget Code]) LIKE '%VALVE%'
OR upper([Budget Code]) LIKE '% WALL%'
OR upper([Budget Code]) LIKE '%WAREHOUSE%'
OR upper([Budget Code]) LIKE '%WATER%MAIN%'
OR upper([Budget Code]) LIKE '%WATER%TOWER%'
OR upper([Budget Code]) LIKE '%WAYFINDING%'
OR upper([Budget Code]) LIKE '%WEIR%'
OR upper([Budget Code]) LIKE '%WETLAND%'
OR upper([Budget Code]) LIKE '%WIDENING%'
OR upper([Budget Code]) LIKE '%WIER%'
OR upper([Budget Code]) LIKE '%WING%'
OR upper([Budget Code]) LIKE '%WORKSPACE%'
OR upper([Budget Code]) LIKE '% WMS %'
OR upper([Budget Code]) LIKE '%WM%REPL%'
OR upper([Budget Code]) LIKE '%YACHT%CLUB%'
OR upper([Budget Code]) LIKE '%YARD%'
OR upper([Budget Code]) LIKE '%YMCA%'
OR upper([Budget Code]) LIKE '%YWHA%'
OR upper([Budget Code]) LIKE '%ZOO%'
)
AND bc_category IS NULL;
"""

# dpr specific -- debug 
query_dpr_fixed_asset = """
UPDATE capital_projects SET bc_category = 'Fixed Asset'
WHERE [Budget Code] REGEXP '[BMQRX][0-9][0-9][0-9]' AND [Agency] = 'Department of Parks and Recreation'
AND bc_category IS NULL;
"""

In [14]:
# --- execute SQL queries on Budget Code to assign bc_category

queries = [query_itt_vehicles_equipment, query_lump_sum, query_fixed_asset, query_dpr_fixed_asset]
replace = {'Budget Code': 'Contract Purpose', 'bc_category':'cp_category'}
for query in queries:
    session.execute(query)
    for bc, cp in replace.items():
        query = query.replace(bc, cp)
    session.execute(query)

# -- execute SQL queries on Contract Purpose to assign cp_category 
# Commit the changes to the database
session.commit()

df_bc_cp_cats = pd.read_sql_table('capital_projects', engine)
engine.dispose()

## Merge Checkbook NYC onto CPDB geometries <a class="anchor" id="fourth-bullet"></a>

Notes:
- 16,687 unique projects ('FMS ID')
- 3,880 joined to geometries from CPDB

In [15]:
df_bc_cp_cats.shape

(16687, 7)

In [16]:
# -- join checkbook nyc to geometries and clean
checkbook_gdf = join_checkbook_geoms(df_bc_cp_cats, all_cpdb)
checkbook_gdf.rename(columns={'typecatego':'cpdb_category'}, inplace=True)
checkbook_geoms_only = checkbook_gdf[checkbook_gdf['_merge']=='both']
checkbook_geoms_only.drop('_merge', axis=1, inplace=True)
checkbook_geoms_only = checkbook_geoms_only[['FMS ID', 'Check Amount', 'Contract Purpose', 'Budget Code', 
'Agency','bc_category', 'cp_category','cpdb_category',
        'ccpversion', 'maprojid', 'magency',
       'magencyacr', 'projectid', 'descriptio',  'geomsource',
       'dataname', 'datasource', 'datadate', 'geometry', 'cartodb_id']]

/var/folders/7p/9414x_x93rvb4p6kc2ny_wl00000gn/T/ipykernel_77067/807676933.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  checkbook_geoms_only.drop('_merge', axis=1, inplace=True)


## Category assignment using high-sensitivity Fixed Asset method

In [23]:
## Category assignment using high-sensitivity Fixed Asset method
# vals
cols = ['cpdb_category', 'bc_category', 'cp_category']
vals = checkbook_geoms_only['cpdb_category'].unique()

def assign_category(row):
    #cols = ['cpdb_category', 'bc_category', 'cp_category']
    cats = ['Fixed Asset', 'ITT, Vehicles, and Equipment', 'Lump Sum', 'None']
    
    if cats[0] in row.values: # <------ can I slice this for performance? or is the difference in cols being iterated over negligible? 
        return cats[0]
    elif cats[1] in row.values:
        return cats[1]
    elif cats[2] in row.values:
        return cats[2]
    else:
        return cats[3]

checkbook_geoms_only['final_category'] = checkbook_geoms_only.apply(lambda row: assign_category(row), axis=1)
checkbook_geoms_only = checkbook_geoms_only[['FMS ID', 'Check Amount', 'Contract Purpose', 'Budget Code', 
'Agency','bc_category', 'cp_category','cpdb_category', 'final_category',
        'ccpversion', 'maprojid', 'magency',
       'magencyacr', 'projectid', 'descriptio',  'geomsource',
       'dataname', 'datasource', 'datadate', 'geometry', 'cartodb_id']]